### PR TITLE
163283560 detected route changes migration

### PR DESCRIPTION
**Pull request by philosotractor**
# Datamodel
* gtfs-transit-changes route-changes renamed to route-changes-old
* new table detected-route-change
* old route-changes migrated to detected-route-change
* composite type gtfs-route-change-info altered

# Changed
* Changed views, controllers and service to utilized new table

### CODE REVIEW

File: database/src/main/resources/db/migration/V1_133__detected_route_changes_table.sql [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+ "route-long-name"            TEXT NOT NULL,          -- e.g. "Helsinki-Oulu"
+ "trip-headsign"              TEXT,
+ "change-type"                "gtfs-route-change-type", 
+ "added-trips"                INTEGER DEFAULT 0,
+ "removed-trips"              INTEGER DEFAULT 0,
+ "trip-stop-sequence-changes-lower"   INTEGER DEFAULT 0 NOT NULL, -- Range of change counts in trips (0 - 2), low limit
+ "trip-stop-sequence-changes-upper"   INTEGER DEFAULT 0 NOT NULL, -- Range of change counts in trips (0 - 2), high limit
+ "trip-stop-time-changes-lower"       INTEGER DEFAULT 0 NOT NULL, -- Range of stop time changes in trips (0 - 200), low limit
+ "trip-stop-time-changes-upper"       INTEGER DEFAULT 0 NOT NULL, -- Range of stop time changes in trips (0 - 200), high limit
+ "current-week-date"          DATE,                   -- Day of detection run
```
Comment (rckopteri) [inline database/src/main/resources/db/migration/V1_133__detected_route_changes_table.sql]: Just making notes: We should change this column name when we change that detection algorithm.

File: database/src/main/resources/db/migration/V1_133__detected_route_changes_table.sql [hunk 2]
```diff
@@ -0,0 +1,91 @@
+CREATE TABLE "detected-route-change" (
```
Comment (rckopteri) [inline database/src/main/resources/db/migration/V1_133__detected_route_changes_table.sql]: btw: you need to update this filename

File: database/src/main/resources/db/migration/V1_133__detected_route_changes_table.sql [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+          -- raise notice '      *************** skipped id=% date=%', r.tsid, r.d;
+        END IF;
+      END LOOP;
+  END
+  $$ LANGUAGE plpgsql;
+
+ALTER TABLE "gtfs-transit-changes"
+  RENAME COLUMN "route-changes" TO "route-changes-old";
+
+ALTER TYPE "gtfs-route-change-info"
```
Comment (rckopteri) [inline database/src/main/resources/db/migration/V1_133__detected_route_changes_table.sql]: You could use only one alter type gtfs-route-change-info and put all thos add attributes to that

File: ote/src/clj/ote/services/transit_visualization.clj [hunk 4]
```diff
@@ -107,18 +109,22 @@
...
-                                                   {:gtfs/transport-service-id service-id}))
-          :gtfs-package-info (fetch-gtfs-packages-for-service db {:service-id service-id})}))
+    (let [service-id (Long/parseLong service-id)]
+      {:service-info (first (fetch-service-info db {:service-id service-id}))
+       :changes (first (detected-changes-for-date db
+                                                  {:service-id service-id
+                                                   :date (-> date
+                                                             time/parse-date-iso-8601
+                                                             java.sql.Date/valueOf)}))
+       :route-changes (detected-route-changes-for-date db
```
Comment (rckopteri) [inline ote/src/clj/ote/services/transit_visualization.clj]: What do you think. Should we have str-date->sql-date function, that would transform string date (2019-01-01) to sql date.

File: ote/src/clj/ote/services/transit_visualization.sql [hunk 5]
```diff
@@ -170,3 +170,22 @@ SELECT p.id, p.created,
   JOIN "external-interface-description" eid ON p."external-interface-description-id" = eid.id
  WHERE p."transport-service-id" = :service-id AND p."deleted?" = FALSE
  ORDER BY p.created DESC;
+
+-- name: detected-changes-for-date
```
Comment (rckopteri) [inline ote/src/clj/ote/services/transit_visualization.sql]: Better name for the query: detected-changed-to-service-by-date

File: ote/src/clj/ote/services/transit_visualization.sql [hunk 6]
```diff
@@ -170,3 +170,22 @@ SELECT p.id, p.created,
...
+SELECT c.date, c."added-routes", c."removed-routes", c."changed-routes", c."no-traffic-routes", c."current-week-date",
+       c."different-week-date", c."change-date", c.created, c."transport-service-id"
+FROM "gtfs-transit-changes" c
+WHERE c."transport-service-id" = :service-id
+  AND c.date = :date::DATE;
+
+-- name: detected-route-changes-for-date
+SELECT r."route-short-name", r."route-long-name", r."trip-headsign",
```
Comment (rckopteri) [inline ote/src/clj/ote/services/transit_visualization.sql]: Better name: changed-routes-to-service-by-date

File: ote/src/clj/ote/transit_changes/detection.clj [hunk 7]
```diff
@@ -25,7 +26,7 @@
                                            {:specql.core/order-by        :gtfs/id
                                             :specql.core/order-direction :desc
                                             :specql.core/limit           package-count}))
-        id-vector (into [] (sort < id-map))]
+        id-vector (vec (sort < id-map))]
```
Comment (rckopteri) [inline ote/src/clj/ote/transit_changes/detection.clj]: peukku

File: database/src/main/resources/db/migration/V1_133__detected_route_changes_table.sql (hunk 2 above)
Comment (philosotractor) [reply database/src/main/resources/db/migration/V1_133__detected_route_changes_table.sql]: Roger

File: database/src/main/resources/db/migration/V1_133__detected_route_changes_table.sql (hunk 3 above)
Comment (philosotractor) [reply database/src/main/resources/db/migration/V1_133__detected_route_changes_table.sql]: Will remove all composite type ALTERs

File: ote/src/clj/ote/services/transit_visualization.sql (hunk 5 above)
Comment (philosotractor) [reply ote/src/clj/ote/services/transit_visualization.sql]: Roger

File: ote/src/clj/ote/transit_changes/detection.clj (hunk 7 above)
Comment (philosotractor) [reply ote/src/clj/ote/transit_changes/detection.clj]: Thanx

File: ote/src/clj/ote/services/transit_visualization.sql (hunk 6 above)
Comment (philosotractor) [reply ote/src/clj/ote/services/transit_visualization.sql]: Roger

File: database/src/main/resources/db/migration/V1_133__detected_route_changes_table.sql [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+  ADD ATTRIBUTE "trip-stop-time-changes-upper" INTEGER;
+
+ALTER TYPE "gtfs-route-change-info"
+  DROP ATTRIBUTE "trip-stop-sequence-changes";
+
+ALTER TYPE "gtfs-route-change-info"
+  DROP ATTRIBUTE "trip-stop-time-changes";
+
+-- TODO: Add an index to this table to date and service-id
```
Comment (erno) [inline database/src/main/resources/db/migration/V1_133__detected_route_changes_table.sql]: Any perf observations from testing with prod data without indexes? Does observed behaviour show that that indexes are needed or not needed?

File: ote/src/clj/ote/services/transit_visualization.sql [hunk 9]
```diff
@@ -170,3 +170,22 @@ SELECT p.id, p.created,
...
+       r."change-type", r."added-trips", r."removed-trips",
+       r."trip-stop-sequence-changes-lower", r."trip-stop-sequence-changes-upper",
+       r."trip-stop-time-changes-lower", r."trip-stop-time-changes-upper",
+       r."current-week-date", r."different-week-date", r."change-date", r."route-hash-id"
+ FROM "detected-route-change" r
+ WHERE r."transit-service-id" = :service-id
+   AND r."transit-change-date" = :date::DATE;
+-- GROUP BY c.date, c."added-routes", c."removed-routes", c."changed-routes", c."no-traffic-routes", c."current-week-date",
```
Comment (erno) [inline ote/src/clj/ote/services/transit_visualization.sql]: delete if unneeded?

File: ote/src/clj/ote/transit_changes/detection.clj [hunk 10]
```diff
@@ -470,23 +474,39 @@
...
+                           (let [update-count (specql/update! db :gtfs/detected-route-change
+                                                              (merge {:gtfs/transit-change-date (sql-date today)
+                                                                      :gtfs/transit-service-id service-id
+                                                                      :gtfs/created-date (java.util.Date.)}
+                                                                     r)
+                                                              {:gtfs/transit-change-date (sql-date today)
+                                                               :gtfs/transit-service-id service-id
+                                                               :gtfs/route-hash-id (:gtfs/route-hash-id r)})]
+
+                             (when (not= 1 update-count)
```
Comment (erno) [inline ote/src/clj/ote/transit_changes/detection.clj]: for future reference, specql also supports upsert - but no need to change now as the code is tested & working.

File: ote/src/cljs/ote/views/transit_visualization.cljs [hunk 11]
```diff
@@ -333,43 +333,39 @@
       ^{:key label}
       [labeled-icon (stylefy/use-style style/transit-changes-legend-icon) [icon] label]))])
 
-(defn format-range [{:keys [lower upper lower-inclusive? upper-inclusive?]}]
+
+(defn format-range [lower upper]
```
Comment (erno) [inline ote/src/cljs/ote/views/transit_visualization.cljs]: Is this function meant to handle a lower=nil upper=non-nil case (format-range nil 42) produces "—2"

Comment (philosotractor) [reply ote/src/cljs/ote/views/transit_visualization.cljs]: True, was wonderingthe same during implementation but apparently even old code didn't handle that scenario. Logic on backend side sets lower boundary to same as upper if it's missing.
Nevertheless I'll consider improving that in the future.

File: ote/src/clj/ote/services/transit_visualization.sql (hunk 9 above)
Comment (philosotractor) [reply ote/src/clj/ote/services/transit_visualization.sql]: Roger

File: ote/src/clj/ote/services/transit_visualization.clj (hunk 4 above)
Comment (philosotractor) [reply ote/src/clj/ote/services/transit_visualization.clj]: Implemented in next PR

File: ote/src/clj/ote/transit_changes/detection.clj (hunk 10 above)
Comment (philosotractor) [reply ote/src/clj/ote/transit_changes/detection.clj]: There was a reason why this was done as update&insert pair...some problem when utilizing upsert.

**Comment by philosotractor** [comment]
Thank you for comments, implemented in https://github.com/finnishtransportagency/mmtis-national-access-point/pull/885
Closing this PR.

**Pull request closed without merging**